beacon-chain/sync/data_columns_reconstruct.go

```diff
@@ -3,210 +3,139 @@ package sync
 		}
 
-		if err := s.broadcastMissingDataColumnSidecars(slot, proposerIndex, root, timeIntoSlot); err != nil {
-			log.WithError(err).Error("Failed to broadcast missing data column sidecars")
-		}
-	})
+		// Randomly choose value before starting reconstruction.
+		randFloat := s.reconstructionRandGen.Float64()
+		timeIntoSlot := time.Duration(maxReconstructionDelaySec*randFloat) * time.Second
+		broadcastTime := slotStartTime.Add(timeIntoSlot)
+		waitingTime := time.Until(broadcastTime)
+
+		wg.Add(1)
+		time.AfterFunc(waitingTime, func() {
+			defer wg.Done()
+
+			// Return early if the context was canceled during the waiting time.
+			if err := ctx.Err(); err != nil {
+				return
+			}
+
+			// Return early if reconstruction is not needed anymore.
+			if !s.shouldReconstruct(root) {
+				return
+			}
+
+			// Load all the stored data column sidecars for this root.
+			verifiedSidecars, err := s.cfg.dataColumnStorage.Get(root, nil)
+			if err != nil {
+				log.WithError(err).Error("Failed to get data column sidecars")
+				return
+			}
+
+			// Reconstruct all the data column sidecars.
+			startTime := time.Now()
+			reconstructedSidecars, err := peerdas.ReconstructDataColumnSidecars(verifiedSidecars)
+			if err != nil {
+				log.WithError(err).Error("Failed to reconstruct data column sidecars")
+				return
+			}
+
+			dataColumnReconstructionHistogram.Observe(float64(time.Since(startTime).Milliseconds()))
+			dataColumnReconstructionCounter.Add(float64(len(reconstructedSidecars) - len(verifiedSidecars)))
+
+			// Retrieve indices of data column sidecars to sample.
+			columnIndicesToSample, err := s.columnIndicesToSample()
+			if err != nil {
+				log.WithError(err).Error("Failed to get column indices to sample")
+				return
+			}
+
+			// Compute sidecars we need to broadcast and receive.
+			unseenSidecars := make([]blocks.VerifiedRODataColumn, 0, len(reconstructedSidecars))
+			for _, sidecar := range reconstructedSidecars {
+				// Skip already seen data column sidecars.
+				if s.hasSeenDataColumnIndex(slot, proposerIndex, sidecar.Index) {
+					continue
+				}
+
+				if columnIndicesToSample[sidecar.Index] {
+					unseenSidecars = append(unseenSidecars, sidecar)
+				}
+			}
+
+			// Broadcast all the data column sidecars we reconstructed but did not see via gossip.
+			for _, sidecar := range unseenSidecars {
+				// Compute the subnet for this data column sidecar.
+				subnet := peerdas.ComputeSubnetForDataColumnSidecar(sidecar.Index)
+
+				// Broadcast the data column sidecar.
+				if err := s.cfg.p2p.BroadcastDataColumnSidecar(subnet, sidecar); err != nil {
+					log.WithError(err).Error("Broadcast data column")
+				}
+			}
+
+			// Receive data column sidecars.
+			if err := s.receiveDataColumnSidecars(ctx, unseenSidecars); err != nil {
+				log.WithError(err).Error("Failed to receive data column sidecars")
+				return
+			}
+		})
+
+		wg.Wait()
+		return nil, nil
+	}); err != nil {
+		return err
```

Comment:
would it be benefitial to wrap this error with a description that something timed out?